### PR TITLE
ch-image: add history metadata

### LINKED ISSUE #1002: ch-image push: fails for Quay.io
Pushing an image to Quay.io causes HTTP 500 while uploading manifest:

```
$ ch-image push 00_tiny quay.io/charliecloud/test:latest
pushing image:   00_tiny
destination:     quay.io/charliecloud/test:latest
layer 1/1: gathering
layer 1/1: preparing
preparing metadata
starting upload
layer 1/1: 0f73d3d: checking if already in repository
layer 1/1: 0f73d3d: already present
config: 0db8e4b: checking if already in repository
config: 0db8e4b: not present, uploading

Username: reidpr
Password:
manifest: uploading
error: PUT failed; expected status {201, 401} but got 500: Internal Server Error
```

### PR DESCRIPTION
**Pull request by j-ogas**
Closes #1002.

### CODE REVIEW

File: test/build/50_ch-image.bats
Comment (reidpr) [inline test/build/50_ch-image.bats:459]: Shouldn't the last one be `"empty_layer": "False"`?

File: test/build/50_ch-image.bats
Comment (reidpr) [inline test/build/50_ch-image.bats:414]: Do we want the string "True" or the boolean value true?

**Review by reidpr — COMMENTED** [review]
Good progress. Some comments in-line. Two general notes:

* Use case 1: We still flatten the image, so don't we need to modify the image so we don't get a layer count mismatch?

* Rejection 3: Charliecloud treats a multi-stage image as multiple independent builds that happen to be in the same Dockerfile. So I don't think there's a special case here.

File: test/build/50_ch-image.bats
Comment (j-ogas) [reply test/build/50_ch-image.bats]: I haven't found a good way to tell if `image_i` is the last image. So far we've been deferring to `push.py` to append that final non-empty entry.

Comment (j-ogas) [reply test/build/50_ch-image.bats]: Good catch.

**Comment by reidpr** [comment]
FWIW at present this PR is incompatible with pre-PR storage directories:

```
$ ch-test -b ch-image --pack-fmt=tar-unpack all
[...]
executing build phase tests ...
 ✗ ch-build 00_tiny
   (in test file /tmp/ch-test.tmp.reidpr/build_auto.bats, line 13)
     `ch-build $force -t 00_tiny --file="[...]/charliecloud/test/Dockerfile.00_tiny" "[...]/charliecloud/test"' failed
   building with: ch-image
   warning: not yet supported, ignored: issue #777: .dockerignore file
     5 FROM alpine:3.9
   warning: --force not available (no suitable config found)
   Traceback (most recent call last):
     File "[...]/charliecloud/bin/ch-image", line 240, in <module>
       main()
     File "[...]/charliecloud/bin/ch-image", line 234, in main
       cli.func(cli)
     File "[...]/charliecloud/bin/../lib/charliecloud/build.py", line 181, in main
       ml.visit_topdown(tree)
     File "[...]/charliecloud/bin/../lib/charliecloud/lark/visitors.py", line 269, in visit_topdown
       self._call_userfunc(subtree)
     File "[...]/charliecloud/bin/../lib/charliecloud/lark/visitors.py", line 241, in _call_userfunc
       return getattr(self, tree.data, self.__default__)(tree)
     File "[...]/charliecloud/bin/../lib/charliecloud/build.py", line 224, in __default__
       inst.metadata_history_update(self.instruction_ct)
     File "[...]/charliecloud/bin/../lib/charliecloud/build.py", line 278, in metadata_history_update
       hist = images[image_i].metadata_history_return()
     File "[...]/charliecloud/bin/../lib/charliecloud/charliecloud.py", line 490, in metadata_history_return
       hist = self.metadata["history"]
   KeyError: 'history'
```

This should be fixed. I think likely it can be made to work with no storage format version bump.

File: lib/build.py
Comment (reidpr) [inline lib/build.py:224]: This would be better in `execute()` I think. Then it wouldn't need to care about the instruction index.

File: lib/build.py
Comment (reidpr) [inline lib/build.py:281]: These can probably be eliminated in favor of calling the `Image` methods directly.

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:460]: This can probably be eliminated by leaving the `empty_layer` workaround until push time.

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:453]: I'd suggest an explicit translation step here. I.e., do we really want all fields in each history entry from the config? Which ones do we care about and which do we not?

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:478]: Not clear to me what this function does. Looks like a dictionary merge of some kind? If so, there's idiomatic ways to do it w/o defining a function.

Also, if the translation from OCI config moves to pull, this won't need to care about config keys being present.

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:484]: This translation to OCI config should happen in `push.py` I'd think.

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:487]: Why a function instead of just assigning the dictionary key?

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:494]: I'm seeing a lot of these. I would recommend moving the workaround for missing history to when `metadata.json` is read from disk. Coupled with the translate on pull/push, then all the internal history manipulation code can assume the data structures are consistent.

File: lib/build.py
Comment (reidpr) [inline lib/build.py:272]: I'd recommend that the function that adds a history entry take the two? history fields directly, with names chosen by you (IMO `created` and `created_by` are confusing) e.g. `history_append(when, what)`.

File: lib/push.py
Comment (reidpr) [inline lib/push.py:118]: This should IMO become an explicit translation as well.

**Review by reidpr — CHANGES_REQUESTED** [review]
Nice progress and thanks for your patience on my review.

The main suggestion I have is to figure out a history format that's convenient to use, independent of the OCI config, and use that format internally. Factors to consider include (1) what's important to us and (2) what's needed to generate an OCI config that satisfies the registries. Then, translate from OCI config to our format at pull time, and back to OCI config at push time. This will eliminate the need for anything other than the translation code to care about stuff like the `empty_layer` hack.

File: lib/charliecloud.py
Comment (j-ogas) [reply lib/charliecloud.py:453]: @heasterday argues that something generic like `FROM TAG` in leou of the base image's actual history isn't very helpful; the tag TAG referenced could be from a locally built image and unavailable for inspection from a registry like Dockerhub.

Determining what is useful and what isn't seems strange for the base image. I personally prefer to replace it with something simple like `created_by: FROM TAG` OR pass all the base image history unedited aside from the `empty_layer` hack.

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:454]: 1. Wouldn't this be `KeyError`? Regardless I don't personally like the "ask forgiveness" idiom of Python, preferring instead the "ask permission" model. The reasoning is that it can be tricky to know exactly where somewhat generic exceptions like `KeyError` come from, so you might be responding to a different problem than you expected (though not in this case I don't think).

2. Does the exception ever happen? Can we treat a config missing `history` key as malformed?

File: lib/build.py
Comment (reidpr) [inline lib/build.py:265]: Should this be under the `if`?

File: lib/charliecloud.py
Comment (reidpr) [inline lib/charliecloud.py:469]: Couple suggestions here:

1. Call it `history_append`?
2. Why does it take a list of items? It's only called once with a singleton.
3. I'd suggest more structured arguments, rather than arbitrary objects appended to `self.metadata["history"]`. E.g., the call I see could just pass the string that ends up in `created_by`.

File: lib/push.py
Comment (reidpr) [inline lib/push.py:125]: Would be good I think to document what we're doing with history. For example, it looks like we're using the OCI format. Why? What alternatives were considered and rejected?

**Review by reidpr — CHANGES_REQUESTED** [review]
Good progress. Some food for thought in-line.

File: lib/charliecloud.py
Comment (j-ogas) [reply lib/charliecloud.py:469]: > Call it history_append?

I don't really like the huge name `metadata_history_append` either, however, it allows us to group it alphabetically in charliecloud.py with other related metadata functions.

> Why does it take a list of items? It's only called once with a singleton.

Good catch. I'll update it to take a single dictionary.

**Review by reidpr — APPROVED** [review]
Great, thanks. If the tidy looks OK to you and the tests pass, I'll merge.

File: lib/push.py
Comment (j-ogas) [inline lib/push.py:119]: If the last history entry happens to have the `empty_layer` present then we can hit a corner case where will have a layerless image. IIRC the nginx image history has a final entry with an empty layer. So if we built an image with something silly like:

```
FROM nginix
```

We may bump into this issue. That is why I popped the last entry instead of iterating to n -1. Perhaps there is a better way to handle this?

**Review by j-ogas — COMMENTED** [review]
Looks good except for the layerless edge case. See comment.

**Review by reidpr — COMMENTED** [review]
Looks like there's a comment that needs to be resolved.

File: lib/push.py
Comment (reidpr) [inline lib/push.py:135]: Is it really a string? Below it's a boolean.

```suggestion
                 and not hist[i]["empty_layer"])):
```

**Review by reidpr — COMMENTED** [review]
One question in-line.

**Comment by heasterday** [comment]
@j-ogas This looks good to me. I would like to test it but the infrastructure that I was using to spin up the test Quay instance appears to be down. If you are alright with waiting until next week to merge I can test then.

**Comment by reidpr** [comment]
Probably worth waiting IMO since pushing to Quay was the key motivation.

File: test/build/50_ch-image.bats
Comment (heasterday) [inline test/build/50_ch-image.bats:481]: I'm not familiar with `/bin/ash` what shell is this?

File: test/build/50_ch-image.bats
Comment (heasterday) [inline test/build/50_ch-image.bats:514]: While it doesn't appear to cause issues I notice a difference between the `created_by` lines between Docker and Charliecloud:
Docker follows a format like this:
`/bin/sh -c  #(nop) RUN echo hello`
Charliecloud:
`"RUN ['bin/sh', '-c', 'echo hello']"`.  

This difference is intended, correct?

**Comment by heasterday** [comment]
This appears to resolve the issue but I do have a couple minor questions above.

File: test/build/50_ch-image.bats
Comment (reidpr) [reply test/build/50_ch-image.bats:514]: Yes.

File: test/build/50_ch-image.bats
Comment (reidpr) [reply test/build/50_ch-image.bats:481]: It's a bare-bones shell aimed at POSIX compliance with not much else. https://en.wikipedia.org/wiki/Almquist_shell

**Review by heasterday — APPROVED** [review]

**Pull request merged**
